File: homework/thinhbka/lesson5-graph/101. Symmetric Tree.py
LEFT = 0
RIGHT = 1
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        queue = [root]
        while queue:
            next = []
            for i in range(len(queue)):
                node = queue[i]
                other_node = queue[len(queue)-1-i]
                if node:
                    if other_node is None:
                        return False
                    if node.val !=other_node.val:
                        return False
                if node is None:
                    next.append(None)
                    next.append(None)
                else:
                    next.append(node.left)
                    next.append(node.right)
            if all([x for x in next if x is not None]):
                queue = next
            else:
                queue = []
        return True

File: homework/Truong_Duong/Lesson_7/101_Symmetric_Tree.js
// Solution for https://leetcode.com/problems/symmetric-tree/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */

const isSymmetricUtil = (left, right) => {
  if (left == null || right == null) return left == right;
  if (left.val != right.val) return false;

  return isSymmetricUtil(left.left, right.right) && isSymmetricUtil(left.right, right.left);
}

var isSymmetric = function (root) {
  return isSymmetricUtil(root.left, root.right);
};